File: reid-service/src/reid_engine.py
"""
ReID Engine - Core person re-identification using ONNX

Uses OSNet AIN model to generate 512-dimensional embeddings
for person matching across cameras.
"""

import os
import time
import numpy as np
import onnxruntime as ort
from PIL import Image
import io
from typing import Dict, Optional, Tuple, Any
from collections import OrderedDict

FEATURE_DIM = 512
INPUT_HEIGHT = 256
INPUT_WIDTH = 128

# ImageNet normalization constants
IMAGENET_MEAN = np.array([0.485, 0.456, 0.406])
IMAGENET_STD = np.array([0.229, 0.224, 0.225])

class TrackedPerson:
    def __init__(self, person_id: str, embedding: np.ndarray, camera_id: str,
                 camera_name: str, snapshot: bytes = None):
        self.person_id = person_id
        self.first_seen = time.time()
        self.last_seen = time.time()
        self.embedding = embedding
        self.camera_id = camera_id
        self.camera_name = camera_name
        self.snapshot = snapshot

class ReIDEngine:
    def __init__(self):
        self.session: Optional[ort.InferenceSession] = None
        self.tracked_persons: OrderedDict[str, TrackedPerson] = OrderedDict()
        self.tracking_window_ms = 60000  # 60 seconds default
        self.debug_mode = False
        self.similarity_threshold = 0.6  # Minimum similarity to consider same person

    async def initialize(self):
        """Initialize ONNX model"""
        model_path = os.path.join(os.path.dirname(__file__), 'osnet_ain_multisource.onnx')

        try:
            self.session = ort.InferenceSession(model_path)
            self.log('ReID engine initialized with OSNet ONNX model')
        except Exception as e:
            raise Exception(f'Failed to load ONNX model: {e}')

    async def process_detection(self, image_buffer: bytes, camera_id: str,
                               camera_name: str, detection_type: str) -> Dict:
        """
        Process a detection and determine if it's a new person
        Returns: { isNew: bool, personId: str, firstDetection?: TrackedPerson }
        """
        if not self.session:
            raise Exception('ReID engine not initialized')

        # Clean up old entries
        self._cleanup_old_entries()

        # Only process person/face detections for ReID
        if detection_type not in ['person', 'face']:
            # For vehicles/animals, use simple time-based deduplication per camera
            simple_id = f'{detection_type}_{camera_id}_{int(time.time() * 1000 / self.tracking_window_ms)}'
            is_new = simple_id not in self.tracked_persons
            if is_new:
                # Create a dummy entry to track it
                self.tracked_persons[simple_id] = TrackedPerson(
                    simple_id, np.zeros(FEATURE_DIM), camera_id, camera_name
                )
            return {
                'isNew': is_new,
                'personId': simple_id
            }

        # Extract embedding from image
        embedding = await self._extract_embedding(image_buffer)

        # Find best match in tracked persons
        match = self._find_best_match(embedding)

        if match and match[1] >= self.similarity_threshold:
            # Known person - update last seen
            person_id = match[0]
            tracked = self.tracked_persons[person_id]
            tracked.last_seen = time.time()

            self.log(f'ReID match: {person_id} on {camera_name} (similarity: {match[1]:.3f})')

            return {
                'isNew': False,
                'personId': person_id,
                'firstDetection': self._tracked_to_dict(tracked)
            }
        else:
            # New person - create tracking entry
            person_id = self._generate_person_id()

            tracked = TrackedPerson(
                person_id,
                embedding,
                camera_id,
                camera_name,
                image_buffer
            )

            self.tracked_persons[person_id] = tracked

            self.log(f'New person: {person_id} detected on {camera_name}')

            return {
                'isNew': True,
                'personId': person_id,
                'firstDetection': self._tracked_to_dict(tracked)
            }

    async def _extract_embedding(self, image_buffer: bytes) -> np.ndarray:
        """Extract ReID embedding from image"""
        if not self.session:
            raise Exception('Session not initialized')

        # Preprocess image
        tensor = self._preprocess_image(image_buffer)

        # Run inference
        outputs = self.session.run(['output'], {'input': tensor})

        # Get embedding (already L2-normalized by model)
        embedding = outputs[0].squeeze()

        # Verify it's normalized
        norm = np.linalg.norm(embedding)
        if abs(norm - 1.0) > 0.01:
            self.log(f'Warning: Embedding norm is {norm}, expected ~1.0')

        return embedding

    def _preprocess_image(self, image_buffer: bytes) -> np.ndarray:
        """Preprocess image for ReID model"""
        # Load and resize image
        img = Image.open(io.BytesIO(image_buffer))
        img = img.convert('RGB')
        img = img.resize((INPUT_WIDTH, INPUT_HEIGHT), Image.LANCZOS)

        # Convert to numpy array and normalize
        img_array = np.array(img).astype(np.float32) / 255.0

        # Apply ImageNet normalization
        img_array = (img_array - IMAGENET_MEAN) / IMAGENET_STD

        # Transpose to CHW format
        img_array = img_array.transpose(2, 0, 1)

        # Add batch dimension
        img_array = np.expand_dims(img_array, axis=0)

        return img_array.astype(np.float32)

    def _find_best_match(self, embedding: np.ndarray) -> Optional[Tuple[str, float]]:
        """Find best matching person in cache"""
        best_match = None
        best_similarity = 0

        for person_id, tracked in self.tracked_persons.items():
            # Skip non-person entries
            if not isinstance(tracked.embedding, np.ndarray) or tracked.embedding.shape[0] != FEATURE_DIM:
                continue

            similarity = self._cosine_similarity(embedding, tracked.embedding)

            if similarity > best_similarity:
                best_similarity = similarity
                best_match = (person_id, similarity)

        return best_match

    def _cosine_similarity(self, a: np.ndarray, b: np.ndarray) -> float:
        """Calculate cosine similarity between embeddings"""
        # Since embeddings are L2-normalized, this is just dot product
        return float(np.dot(a, b))

    def _generate_person_id(self) -> str:
        """Generate unique person ID"""
        import random
        import string
        rand_str = ''.join(random.choices(string.ascii_lowercase + string.digits, k=9))
        return f'person_{int(time.time() * 1000)}_{rand_str}'

    def _cleanup_old_entries(self):
        """Remove entries older than tracking window"""
        current_time = time.time()
        cutoff_time = current_time - (self.tracking_window_ms / 1000)

        # Find keys to remove
        keys_to_remove = []
        for person_id, tracked in self.tracked_persons.items():
            if tracked.last_seen < cutoff_time:
                keys_to_remove.append(person_id)

        # Remove old entries
        for key in keys_to_remove:
            del self.tracked_persons[key]

    def _tracked_to_dict(self, tracked: TrackedPerson) -> Dict:
        """Convert TrackedPerson to dictionary"""
        return {
            'personId': tracked.person_id,
            'firstSeen': tracked.first_seen * 1000,  # Convert to ms
            'lastSeen': tracked.last_seen * 1000,
            'cameraId': tracked.camera_id,
            'cameraName': tracked.camera_name,
            'snapshot': tracked.snapshot
        }

    def set_tracking_window(self, ms: int):
        """Set tracking window duration"""
        self.tracking_window_ms = ms
        # Clean up based on new window
        self._cleanup_old_entries()

    def set_debug_mode(self, enabled: bool):
        """Enable/disable debug logging"""
        self.debug_mode = enabled

    def get_stats(self) -> Dict:
        """Get current stats"""
        return {
            'trackedPersons': len(self.tracked_persons),
            'cacheSize': len(self.tracked_persons) * FEATURE_DIM * 4  # Approximate bytes
        }

    def clear(self):
        """Clear all tracked persons"""
        self.tracked_persons.clear()
        self.log('Cleared all tracked persons')

    def log(self, message: str):
        """Log message if debug mode is enabled"""
        if self.debug_mode:
            print(f'[ReID] {message}')

File: reid-service/src_typescript_backup/reid-engine.ts
/**
 * ReID Engine - Core person re-identification using ONNX
 *
 * Uses OSNet AIN model to generate 512-dimensional embeddings
 * for person matching across cameras.
 */

import * as ort from 'onnxruntime-node';
import { LRUCache } from 'lru-cache';
import * as path from 'path';
import sharp from 'sharp';

const FEATURE_DIM = 512;
const INPUT_HEIGHT = 256;
const INPUT_WIDTH = 128;

// ImageNet normalization constants
const IMAGENET_MEAN = [0.485, 0.456, 0.406];
const IMAGENET_STD = [0.229, 0.224, 0.225];

interface TrackedPerson {
    personId: string;
    firstSeen: number;
    lastSeen: number;
    embedding: Float32Array;
    cameraId: string;
    cameraName: string;
    snapshot?: Buffer;  // Original snapshot from first detection
}

export class ReIDEngine {
    private session: ort.InferenceSession | null = null;
    private trackedPersons: LRUCache<string, TrackedPerson>;
    private trackingWindowMs = 60000; // 60 seconds default
    private debugMode = false;
    private similarityThreshold = 0.6; // Minimum similarity to consider same person

    constructor() {
        // LRU cache automatically expires old entries
        this.trackedPersons = new LRUCache<string, TrackedPerson>({
            max: 1000, // Maximum tracked persons
            ttl: this.trackingWindowMs,
            updateAgeOnGet: false,
            updateAgeOnHas: false
        });
    }

    async initialize(): Promise<void> {
        const modelPath = path.join(__dirname, '..', 'models', 'osnet_ain_multisource.onnx');

        try {
            this.session = await ort.InferenceSession.create(modelPath);
            this.log('ReID engine initialized with OSNet ONNX model');
        } catch (error) {
            throw new Error(`Failed to load ONNX model: ${error}`);
        }
    }

    /**
     * Process a detection and determine if it's a new person
     * @returns { isNew: boolean, personId: string, firstDetection?: TrackedPerson }
     */
    async processDetection(
        imageBuffer: Buffer,
        cameraId: string,
        cameraName: string,
        detectionType: string
    ): Promise<{
        isNew: boolean;
        personId: string;
        firstDetection?: TrackedPerson;
    }> {
        if (!this.session) {
            throw new Error('ReID engine not initialized');
        }

        // Only process person/face detections for ReID
        if (detectionType !== 'person' && detectionType !== 'face') {
            // For vehicles/animals, use simple time-based deduplication per camera
            const simpleId = `${detectionType}_${cameraId}_${Math.floor(Date.now() / this.trackingWindowMs)}`;
            return {
                isNew: !this.trackedPersons.has(simpleId),
                personId: simpleId
            };
        }

        // Extract embedding from image
        const embedding = await this.extractEmbedding(imageBuffer);

        // Find best match in tracked persons
        const match = this.findBestMatch(embedding);

        const now = Date.now();

        if (match && match.similarity >= this.similarityThreshold) {
            // Known person - update last seen
            const tracked = this.trackedPersons.get(match.personId)!;
            tracked.lastSeen = now;

            this.log(`ReID match: ${match.personId} on ${cameraName} (similarity: ${match.similarity.toFixed(3)})`);

            return {
                isNew: false,
                personId: match.personId,
                firstDetection: tracked
            };
        } else {
            // New person - create tracking entry
            const personId = this.generatePersonId();

            const tracked: TrackedPerson = {
                personId,
                firstSeen: now,
                lastSeen: now,
                embedding,
                cameraId,
                cameraName,
                snapshot: imageBuffer
            };

            this.trackedPersons.set(personId, tracked);

            this.log(`New person: ${personId} detected on ${cameraName}`);

            return {
                isNew: true,
                personId,
                firstDetection: tracked
            };
        }
    }

    /**
     * Extract ReID embedding from image
     */
    private async extractEmbedding(imageBuffer: Buffer): Promise<Float32Array> {
        if (!this.session) {
            throw new Error('Session not initialized');
        }

        // Preprocess image: resize to 256x128 and normalize
        const tensor = await this.preprocessImage(imageBuffer);

        // Run inference
        const feeds = { input: tensor };
        const results = await this.session.run(feeds);

        // Get embedding (already L2-normalized by model)
        const output = results.output;
        const embedding = new Float32Array(output.data as Float32Array);

        // Verify it's normalized
        const norm = this.calculateNorm(embedding);
        if (Math.abs(norm - 1.0) > 0.01) {
            this.log(`Warning: Embedding norm is ${norm}, expected ~1.0`);
        }

        return embedding;
    }

    /**
     * Preprocess image for ReID model
     */
    private async preprocessImage(imageBuffer: Buffer): Promise<ort.Tensor> {
        // Resize to 256x128 using sharp
        const resized = await sharp(imageBuffer)
            .resize(INPUT_WIDTH, INPUT_HEIGHT, {
                fit: 'fill',
                kernel: 'lanczos3'
            })
            .removeAlpha()
            .raw()
            .toBuffer();

        // Convert to RGB tensor with ImageNet normalization
        const pixelCount = INPUT_HEIGHT * INPUT_WIDTH;
        const tensorData = new Float32Array(3 * pixelCount);

        // Sharp outputs in RGB order
        for (let i = 0; i < pixelCount; i++) {
            const pixelIndex = i * 3;

            // R channel
            tensorData[0 * pixelCount + i] =
                (resized[pixelIndex] / 255.0 - IMAGENET_MEAN[0]) / IMAGENET_STD[0];

            // G channel
            tensorData[1 * pixelCount + i] =
                (resized[pixelIndex + 1] / 255.0 - IMAGENET_MEAN[1]) / IMAGENET_STD[1];

            // B channel
            tensorData[2 * pixelCount + i] =
                (resized[pixelIndex + 2] / 255.0 - IMAGENET_MEAN[2]) / IMAGENET_STD[2];
        }

        // Create ONNX tensor (batch_size=1, channels=3, height=256, width=128)
        return new ort.Tensor('float32', tensorData, [1, 3, INPUT_HEIGHT, INPUT_WIDTH]);
    }

    /**
     * Find best matching person in cache
     */
    private findBestMatch(embedding: Float32Array): { personId: string; similarity: number } | null {
        let bestMatch: { personId: string; similarity: number } | null = null;
        let bestSimilarity = 0;

        for (const [personId, tracked] of this.trackedPersons.entries()) {
            const similarity = this.cosineSimilarity(embedding, tracked.embedding);

            if (similarity > bestSimilarity) {
                bestSimilarity = similarity;
                bestMatch = { personId, similarity };
            }
        }

        return bestMatch;
    }

    /**
     * Calculate cosine similarity between embeddings
     * Note: Since embeddings are L2-normalized, this is just dot product
     */
    private cosineSimilarity(a: Float32Array, b: Float32Array): number {
        let dotProduct = 0;
        for (let i = 0; i < FEATURE_DIM; i++) {
            dotProduct += a[i] * b[i];
        }
        return dotProduct;
    }

    /**
     * Calculate L2 norm of vector
     */
    private calculateNorm(vec: Float32Array): number {
        let sum = 0;
        for (let i = 0; i < vec.length; i++) {
            sum += vec[i] * vec[i];
        }
        return Math.sqrt(sum);
    }

    /**
     * Generate unique person ID
     */
    private generatePersonId(): string {
        return `person_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    }

    /**
     * Set tracking window duration
     */
    setTrackingWindow(ms: number): void {
        this.trackingWindowMs = ms;
        // Update cache TTL
        this.trackedPersons = new LRUCache<string, TrackedPerson>({
            max: 1000,
            ttl: ms,
            updateAgeOnGet: false,
            updateAgeOnHas: false
        });
    }

    /**
     * Enable/disable debug logging
     */
    setDebugMode(enabled: boolean): void {
        this.debugMode = enabled;
    }

    /**
     * Get current stats
     */
    getStats(): {
        trackedPersons: number;
        cacheSize: number;
    } {
        return {
            trackedPersons: this.trackedPersons.size,
            cacheSize: this.trackedPersons.size * FEATURE_DIM * 4 // Approximate bytes
        };
    }

    /**
     * Clear all tracked persons
     */
    clear(): void {
        this.trackedPersons.clear();
        this.log('Cleared all tracked persons');
    }

    private log(message: string): void {
        if (this.debugMode) {
            console.log(`[ReID] ${message}`);
        }
    }
}